File: Stack/Problems/Infix2Postfix/solution.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>

int prec(char op)
{
    if (op == '$')
    {
        return 3;
    }
    else if (op == '*' || op == '/')
    {
        return 2;
    }
    else if (op == '+' || op == '-')
    {
        return 1;
    }
    else
    {
        return -1;
    }
}

void infixToPostFix(char *exp)
{
    int len = strlen(exp);
    char postfix_string[len + 1];
    int trac = 0;
    char stack[len];
    int top = -1;

    for (int i = 0; i < len; i++)
    {
        char s_c = exp[i];

        if (s_c == '(')
        {
            stack[++top] = s_c;
        }

        else if (s_c == ')')
        {
            while (top != -1 && stack[top] != '(')
            {
                postfix_string[trac++] = stack[top--];
            }
            if (top >= 0 && stack[top] == '(')
            {
                top--; // Pop the '('
            }
        }

        // In case of operand
        else if (isalnum(s_c))
        {
            postfix_string[trac++] = s_c;
        }

        // In case of operator
        else
        {
            while (top != -1 && prec(s_c) <= prec(stack[top]))
            {
                postfix_string[trac++] = stack[top--];
            }

            stack[++top] = s_c;
        }
    }

    while (top != -1)
    {
        postfix_string[trac++] = stack[top--];
    }

    postfix_string[trac] = '\0'; // Null-terminate the string
    printf("Postfix Expression: %s\n", postfix_string);
}

int main()
{
    char infix_exp[] = "A+(B*C-(D/E$F)*G)*H";

    printf("Infix Expression: %s\n", infix_exp);
    infixToPostFix(infix_exp);

    return 0;
}

File: Stack/Problems/Infix2Postfix/solution.py
# Function to return precedence of operators
def prec(c):
    if c == "^" or c == "$":
        return 3
    elif c == "/" or c == "*":
        return 2
    elif c == "+" or c == "-":
        return 1
    else:
        return -1


# Function to perform infix to postfix conversion
def infixToPostfix(s):
    st = []
    result = ""

    for i in range(len(s)):
        c = s[i]

        # If the scanned character is
        # an operand, add it to the output string.
        if (
            (c >= "a" and c <= "z")
            or (c >= "A" and c <= "Z")
            or (c >= "0" and c <= "9")
        ):
            result += c

        # If the scanned character is an
        # "(", push it to the stack
        elif c == "(":
            st.append("(")
        
        elif c == ")":
            while st[-1] != "(":
                result += st.pop()
            
            st.pop()
        
        else: 
            while st and (prec(c) < prec(st[-1]) or prec(c) == prec(st[-1])):
                result += st.pop()
            
            st.append(c)
        
    
    while st:
        result += st.pop()
    
    print(result)


exp = "A+(B*C-(D/E$F)*G)*H"
infixToPostfix(exp)
        
